File: JavaScript-SDK/sdk.js
const axios = require('axios');
const xxhash = require('xxhashjs');
const bls = require('bls.js'); // You might need to find the correct BLS library

// Utility hash function using xxhash
const hash = (input) => xxhash.h64().update(input).digest().toString('hex');

// Function to fetch operators using a GraphQL query
async function getOperators() {
    const subgraphUrl = "https://api.studio.thegraph.com/query/85556/bls_apk_registry/version/latest";
    const query = `
        query MyQuery {
          operators {
            id
            operatorId
            pubkeyG1_X
            pubkeyG1_Y
            pubkeyG2_X
            pubkeyG2_Y
            socket
            stake
          }
        }
    `;
    
    const response = await axios.post(subgraphUrl, { query: query });
    const operators = response.data.data.operators;
    
    operators.forEach((operator) => {
        operator.stake = Math.min(1, parseFloat(operator.stake) / Math.pow(10, 18));
        let publicKeyG2 = `1 ${operator.pubkeyG2_X[1]} ${operator.pubkeyG2_X[0]} ${operator.pubkeyG2_Y[1]} ${operator.pubkeyG2_Y[0]}`;
        operator.public_key_g2 = new bls.G2Point();
        operator.public_key_g2.setStr(publicKeyG2);
    });

    return Object.fromEntries(operators.map(operator => [operator.id, operator]));
}

// Zellular class definition
class Zellular {
    constructor(appName, baseUrl, thresholdPercent = 67) {
        this.appName = appName;
        this.baseUrl = baseUrl;
        this.thresholdPercent = thresholdPercent;
        this.operators = null;
        this.aggregatedPublicKey = new bls.G2Point();

        // Fetch operators and initialize public keys
        getOperators().then(operators => {
            this.operators = operators;
            for (let operator of Object.values(operators)) {
                this.aggregatedPublicKey = this.aggregatedPublicKey.add(operator.public_key_g2);
            }
        });
    }

    // Verify a BLS signature
    verifySignature(message, signatureHex, nonsigners) {
        let totalStake = Object.values(this.operators).reduce((sum, operator) => sum + operator.stake, 0);
        nonsigners = nonsigners.map(id => this.operators[id]);
        let nonsignersStake = nonsigners.reduce((sum, operator) => sum + operator.stake, 0);

        if ((100 * nonsignersStake / totalStake) > (100 - this.thresholdPercent)) {
            return false;
        }

        let publicKey = this.aggregatedPublicKey;
        nonsigners.forEach(operator => {
            publicKey = publicKey.sub(operator.public_key_g2);
        });

        let signature = new bls.Signature();
        signature.setStr(signatureHex);

        message = hash(message);
        return signature.verify(publicKey, message);
    }

    // Verify a finalized batch
    verifyFinalized(data, batchHash, chainingHash) {
        let message = JSON.stringify({
            app_name: this.appName,
            state: 'locked',
            index: data.index,
            hash: batchHash,
            chaining_hash: chainingHash
        });

        let signature = data.finalization_signature;
        let nonsigners = data.nonsigners;
        let result = this.verifySignature(message, signature, nonsigners);

        console.log(`App: ${this.appName}, Index: ${data.index}, Verification Result: ${result}`);
        return result;
    }

    // Fetch finalized batches from the backend
    async getFinalized(after, chainingHash) {
        let res = [];
        let index = chainingHash !== null ? after : after - 1;

        while (true) {
            const resp = await axios.get(`${this.baseUrl}/node/${this.appName}/batches/finalized?after=${index}`);
            const data = resp.data.data;

            if (!data) continue;

            let { batches, finalized } = data;
            if (!chainingHash) {
                chainingHash = data.first_chaining_hash;
                batches = batches.slice(1);
                index += 1;
            }

            for (let batch of batches) {
                index += 1;
                chainingHash = hash(chainingHash + hash(batch));
                res.push(batch);

                if (finalized && index === finalized.index) {
                    const isValid = this.verifyFinalized(finalized, hash(batch), chainingHash);
                    if (!isValid) {
                        throw new Error('Invalid signature');
                    }
                    return { chainingHash, res };
                }
            }
        }
    }

    // Generator to yield batches
    async *batches(after = 0) {
        let chainingHash = after === 0 ? '' : null;

        while (true) {
            let { chainingHash: newChainingHash, res: batches } = await this.getFinalized(after, chainingHash);
            chainingHash = newChainingHash;

            for (let batch of batches) {
                after += 1;
                yield { batch, after };
            }
        }
    }

    // Get the last finalized batch
    async getLastFinalized() {
        const url = `${this.baseUrl}/node/${this.appName}/batches/finalized/last`;
        const resp = await axios.get(url);
        const data = resp.data.data;

        const verified = this.verifyFinalized(data, data.hash, data.chaining_hash);
        if (!verified) {
            throw new Error('Invalid signature');
        }

        return data;
    }

    // Send a batch
    async send(batch, blocking = false) {
        if (blocking) {
            let lastFinalized = await this.getLastFinalized();
            let index = lastFinalized.index;

            for await (let { batch: receivedBatch, after: newIndex } of this.batches(index)) {
                if (batch === JSON.parse(receivedBatch)) {
                    return newIndex;
                }
            }
        }

        const url = `${this.baseUrl}/node/${this.appName}/batches`;
        const resp = await axios.put(url, batch);
        if (resp.status !== 200) {
            throw new Error('Failed to send batch');
        }
    }
}

// Main function to run the example
(async () => {
    const operators = await getOperators();
    const baseUrl = Object.values(operators)[Math.floor(Math.random() * Object.values(operators).length)].socket;

    console.log(baseUrl);

    const verifier = new Zellular('simple_app', baseUrl);
    for await (let { batch, after } of verifier.batches()) {
        const txs = JSON.parse(batch);
        txs.forEach((tx, i) => {
            console.log(after, i, tx);
        });
    }
})();

File: Rust-SDK/sdk.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::time::SystemTime;
use bls12_381::{G2Prepared, G2Affine, pairing};

// Struct for operators
#[derive(Debug, Deserialize)]
struct Operator {
    id: String,
    operator_id: String,
    pubkey_g1_x: Vec<String>,
    pubkey_g1_y: Vec<String>,
    pubkey_g2_x: Vec<String>,
    pubkey_g2_y: Vec<String>,
    socket: String,
    stake: f64,
    public_key_g2: Option<G2Affine>, // Placeholder for G2 affine key
}

// Struct for GraphQL query response
#[derive(Debug, Deserialize)]
struct QueryResponse {
    data: QueryData,
}

#[derive(Debug, Deserialize)]
struct QueryData {
    operators: Vec<Operator>,
}

// Hash function using SHA-256
fn hash(input: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(input.as_bytes());
    format!("{:x}", hasher.finalize())
}

// Fetch operators via a GraphQL query
async fn get_operators(client: &Client) -> Result<HashMap<String, Operator>, Box<dyn std::error::Error>> {
    let query = r#"
        query MyQuery {
          operators {
            id
            operatorId
            pubkeyG1_X
            pubkeyG1_Y
            pubkeyG2_X
            pubkeyG2_Y
            socket
            stake
          }
        }
    "#;

    let subgraph_url = "https://api.studio.thegraph.com/query/85556/bls_apk_registry/version/latest";

    let resp = client
        .post(subgraph_url)
        .json(&serde_json::json!({ "query": query }))
        .send()
        .await?;

    let response_json: QueryResponse = resp.json().await?;
    let mut operators = HashMap::new();

    for mut operator in response_json.data.operators {
        operator.stake = f64::min(1.0, operator.stake / 10f64.powi(18));

        // Here, we should set the G2 key (use a proper BLS library for Rust)
        let public_key_g2 = G2Affine::identity(); // Placeholder
        operator.public_key_g2 = Some(public_key_g2);

        operators.insert(operator.id.clone(), operator);
    }

    Ok(operators)
}

// Struct for Zellular
struct Zellular {
    app_name: String,
    base_url: String,
    threshold_percent: f64,
    operators: HashMap<String, Operator>,
    aggregated_public_key: G2Affine,
}

impl Zellular {
    pub async fn new(app_name: &str, base_url: &str, threshold_percent: f64) -> Result<Self, Box<dyn std::error::Error>> {
        let client = Client::new();
        let operators = get_operators(&client).await?;

        // Aggregate G2 public keys
        let mut aggregated_public_key = G2Affine::identity();
        for operator in operators.values() {
            aggregated_public_key = aggregated_public_key + operator.public_key_g2.unwrap();
        }

        Ok(Self {
            app_name: app_name.to_string(),
            base_url: base_url.to_string(),
            threshold_percent,
            operators,
            aggregated_public_key,
        })
    }

    // Verify a BLS signature (placeholder, adjust with a real BLS library)
    pub fn verify_signature(&self, message: &str, signature_hex: &str, nonsigners: Vec<String>) -> bool {
        let total_stake: f64 = self.operators.values().map(|op| op.stake).sum();
        let nonsigners_stake: f64 = nonsigners.iter().map(|id| self.operators.get(id).unwrap().stake).sum();

        if (100.0 * nonsigners_stake / total_stake) > (100.0 - self.threshold_percent) {
            return false;
        }

        let mut public_key = self.aggregated_public_key;
        for nonsigner in nonsigners {
            public_key = public_key - self.operators.get(&nonsigner).unwrap().public_key_g2.unwrap();
        }

        // Replace with BLS signature decoding and verification
        let message_hash = hash(message);
        let signature = G2Affine::identity(); // Placeholder
        pairing(&G2Prepared::from(public_key), &signature).is_zero() // Adjust this line with the actual BLS verification logic
    }

    // Fetch finalized batches (simplified version)
    pub async fn get_finalized(&self, after: i32, chaining_hash: Option<String>) -> Result<(String, Vec<String>), Box<dyn std::error::Error>> {
        let client = Client::new();
        let mut res = Vec::new();
        let mut index = if chaining_hash.is_some() { after } else { after - 1 };

        loop {
            let url = format!("{}/node/{}/batches/finalized?after={}", self.base_url, self.app_name, index);
            let resp = client.get(&url).send().await?.json::<serde_json::Value>().await?;

            if resp["data"].is_null() {
                continue;
            }

            let batches = resp["data"]["batches"].as_array().unwrap_or(&vec![]);
            let finalized = &resp["data"]["finalized"];

            for batch in batches {
                index += 1;
                let chaining_hash = chaining_hash.clone().unwrap_or_else(|| "".to_string()) + &hash(batch.as_str().unwrap());
                res.push(batch.as_str().unwrap().to_string());

                if finalized != &serde_json::Value::Null && index == finalized["index"].as_i64().unwrap() as i32 {
                    return Ok((chaining_hash, res));
                }
            }
        }
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let client = Client::new();
    let operators = get_operators(&client).await?;
    let base_url = operators.values().next().unwrap().socket.clone();

    println!("Base URL: {}", base_url);

    let verifier = Zellular::new("simple_app", &base_url, 67.0).await?;
    let (chaining_hash, batches) = verifier.get_finalized(0, None).await?;

    for (i, batch) in batches.iter().enumerate() {
        println!("Batch {}: {}", i + 1, batch);
    }

    Ok(())
}